### PR TITLE
rewrite cupyx.scipy.ndimage.interpolation using ElementwiseKernel

### LINKED ISSUE #2607: `cupyx.scipy.ndimage.rotate()` saturates P100's 16G memory for 1G input data
We discovered this out-of-memory (OOM) issue in a recent GPU Hackathon with @jqyin and @dxm447, and I'm revisiting this. 

The implementation along the call path `cupyx.scipy.ndimage.rotate()` -> `affine_transform()` -> `map_coordinates()` is very memory-inefficient. The memory inflation from a 1G input to over 16G on a P100 as shown in the following example (extracted from a real-life workload) is not acceptable.
```python
>>> import cupy as cp
>>> from cupyx.scipy.ndimage import rotate
>>>
>>> a = cp.random.random((512, 512, 512))  # 1GB data
>>> b = rotate(a, 15, axes=(1,2))
/home/leofang/test/cupy2/cupyx/scipy/ndimage/interpolation.py:27: UserWarning: In the current feature the default order of rotate is 1. It is different from scipy.ndimage and can change in the future.
  'the future.'.format(func_name))
/home/leofang/test/cupy2/cupyx/scipy/ndimage/interpolation.py:27: UserWarning: In the current feature the default order of affine_transform is 1. It is different from scipy.ndimage and can change in the future.
  'the future.'.format(func_name))
/home/leofang/test/cupy2/cupyx/scipy/ndimage/interpolation.py:27: UserWarning: In the current feature the default order of map_coordinates is 1. It is different from scipy.ndimage and can change in the future.
  'the future.'.format(func_name))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/leofang/test/cupy2/cupyx/scipy/ndimage/interpolation.py", line 351, in rotate
    mode, cval, prefilter)
  File "/home/leofang/test/cupy2/cupyx/scipy/ndimage/interpolation.py", line 242, in affine_transform
    prefilter).reshape(output_shape)
  File "/home/leofang/test/cupy2/cupyx/scipy/ndimage/interpolation.py", line 112, in map_coordinates
    coordinates_floor = cupy.floor(coordinates).astype(cupy.int32)
  File "cupy/core/core.pyx", line 320, in cupy.core.core.ndarray.astype
  File "cupy/core/core.pyx", line 378, in cupy.core.core.ndarray.astype
  File "cupy/core/core.pyx", line 134, in cupy.core.core.ndarray.__init__
  File "cupy/cuda/memory.pyx", line 540, in cupy.cuda.memory.alloc
  File "cupy/cuda/memory.pyx", line 1186, in cupy.cuda.memory.MemoryPool.malloc
  File "cupy/cuda/memory.pyx", line 1207, in cupy.cuda.memory.MemoryPool.malloc
  File "cupy/cuda/memory.pyx", line 994, in cupy.cuda.memory.SingleDeviceMemoryPool.malloc
  File "cupy/cuda/memory.pyx", line 1009, in cupy.cuda.memory.SingleDeviceMemoryPool._malloc
  File "cupy/cuda/memory.pyx", line 743, in cupy.cuda.memory._try_malloc
cupy.cuda.memory.OutOfMemoryError: Out of memory allocating 2,415,384,576 bytes (allocated so far: 15,566,063,104 bytes).
```
Using `pdb` I was able to spot that the biggest memory consumer is in `affine_transform()`:
https://github.com/cupy/cupy/blob/93547cedd3a87074e42edda5e88922c22e5f04b4/cupyx/scipy/ndimage/interpolation.py#L237
which alone took about 4.5G (=3\*512\*627\*627\*8 bytes). 

Another inefficiency I could tell that contributes to OOM is also in `affine_transform()`: 
https://github.com/cupy/cupy/blob/93547cedd3a87074e42edda5e88922c22e5f04b4/cupyx/scipy/ndimage/interpolation.py#L240-L242
which shows that an output buffer `ret` is allocated but not passed into `map_coordinates()`, so in `map_coordinates()` yet another buffer `ret` is allocated:  
https://github.com/cupy/cupy/blob/93547cedd3a87074e42edda5e88922c22e5f04b4/cupyx/scipy/ndimage/interpolation.py#L91
But even if this is fixed, I still got OOM in `map_coordinates()` here:
https://github.com/cupy/cupy/blob/93547cedd3a87074e42edda5e88922c22e5f04b4/cupyx/scipy/ndimage/interpolation.py#L125 
At this point I am not sure what's the best strategy to proceed, so I thought it's better to pause and report back.

Our current workaround is to batch process the images along the first axis, which is OK but less efficient.

### PR DESCRIPTION
**Pull request by grlee77**
This PR improves the performance and reduces memory usage for the `cupyx.scipy.ndimage.interpolation` module. There should be no change in behavior from the prior implementation.

Memory overhead in various functions like `affine_transform` is greatly reduced by not explicitly generating dense n-dimensional coordinate arrays.

closes #2607
also resolves some issues with performance that were previously reported in: #1168, #2345, #2844

Will post the benchmarks in a follow-up comment, but performance is greatly improved on all cases I tested.

The main downside to this PR is that such template-based code generation is a bit harder to read than the previous implementations, but I tried to add some comments describing what is going on.

### CODE REVIEW

**Comment by grlee77** [comment]
The following timings were measured for this PR vs. current master (8.0.0a1) for various operations applied to a 2D float32 image of shape (4096, 4096).

The case labeled `affine_transform_1dmat` corresponds to a different code path in `affine_transform` for the case when the matrix is 1d (corresponding to a zoom).

The next-to-last column gives the acceleration factor for the PR vs. current master, showing from between 4x and 26x improvement!


|                                            |   SciPy (1.4.1) |   CuPy (8.0.0a1) |   CuPy (proposed) |   accel. vs. 8.0.0a1 |   accel. vs. SciPy |
|:-------------------------------------------|----------------:|-----------------:|------------------:|--------------------:|-------------------:|
| rotate (constant, order=0)                 |          0.8071 |           0.0327 |            0.0042 |              7.8077 |           192.8056 |
| rotate (nearest, order=0)                  |          1.0220 |           0.0295 |            0.0042 |              7.0459 |           243.6998 |
| rotate (constant, order=1)                 |          1.0053 |           0.1229 |            0.0052 |             23.7408 |           194.2177 |
| rotate (nearest, order=1)                  |          1.5673 |           0.1183 |            0.0069 |             17.1905 |           227.7342 |
| shift (constant, order=0)                  |          0.1681 |           0.0167 |            0.0012 |             13.7808 |           138.9603 |
| shift (nearest, order=0)                   |          0.1661 |           0.0153 |            0.0009 |             17.6277 |           191.7065 |
| shift (constant, order=1)                  |          0.3448 |           0.0411 |            0.0021 |             19.4609 |           163.4667 |
| shift (nearest, order=1)                   |          0.3636 |           0.0396 |            0.0019 |             21.1493 |           194.3028 |
| zoom (constant, order=0)                   |          0.1746 |           0.0164 |            0.0012 |             13.2128 |           140.5852 |
| zoom (nearest, order=0)                    |          0.1673 |           0.0152 |            0.0008 |             18.3022 |           201.7883 |
| zoom (constant, order=1)                   |          0.3637 |           0.0635 |            0.0029 |             21.6818 |           124.1133 |
| zoom (nearest, order=1)                    |          0.3581 |           0.0623 |            0.0027 |             23.0291 |           132.3654 |
| affine_transform (constant, order=0)       |          0.3592 |           0.0174 |            0.0021 |              8.1698 |           168.8159 |
| affine_transform (nearest, order=0)        |          0.6480 |           0.0152 |            0.0023 |              6.5670 |           279.7564 |
| affine_transform (constant, order=1)       |          0.4340 |           0.0664 |            0.0025 |             26.0903 |           170.6035 |
| affine_transform (nearest, order=1)        |          0.9803 |           0.0626 |            0.0036 |             17.5132 |           274.3002 |
| map_coordinates (constant, order=0)        |          0.4296 |           0.0106 |            0.0022 |              4.7879 |           194.5049 |
| map_coordinates (nearest, order=0)         |          0.4323 |           0.0091 |            0.0021 |              4.2682 |           201.6946 |
| map_coordinates (constant, order=1)        |          0.6561 |           0.0577 |            0.0035 |             16.6911 |           189.9143 |
| map_coordinates (nearest, order=1)         |          0.6641 |           0.0564 |            0.0033 |             16.9896 |           199.9157 |
| affine_transform_1dmat (constant, order=0) |          0.1360 |           0.0167 |            0.0014 |             12.1260 |            98.6344 |
| affine_transform_1dmat (nearest, order=0)  |          0.1680 |           0.0151 |            0.0012 |             12.8153 |           142.2724 |
| affine_transform_1dmat (constant, order=1) |          0.2351 |           0.0487 |            0.0018 |             26.6065 |           128.5304 |
| affine_transform_1dmat (nearest, order=1)  |          0.4457 |           0.0471 |            0.0022 |             21.8071 |           206.5435 |

**Comment by leofang** [comment]
Thanks a lot, Gregory! This is excellent work! I confirmed that this PR eliminates the excessive memory usage we reported in #2607. 

cc: @jqyin @dxm447

**Comment by dxm447** [comment]
Really great!

**Comment by grlee77** [comment]
I suppose the same type of indexing changes made in #3179 should also help here. I can take a look at that when I get a chance.

File: cupyx/scipy/ndimage/_interp_kernels.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+        mode=mode,
+        cval=cval,
+        order=order,
+        name='zoom',
+        integer_output=integer_output,
+    )
+    return cupy.ElementwiseKernel(in_params, out_params, operation, name)
+
+
+@cupy.util.memoize()
```
Comment (leofang) [inline cupyx/scipy/ndimage/_interp_kernels.py]:
```suggestion
@cupy.util.memoize(for_each_device=True)
```

File: cupyx/scipy/ndimage/_interp_kernels.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+        name,
+        order,
+        mode,
+        '_'.join(['{}'.format(j) for j in xshape]),
+        '_'.join(['{}'.format(j) for j in yshape]),
+    )
+    return operation, name
+
+
+@cupy.util.memoize()
```
Comment (leofang) [inline cupyx/scipy/ndimage/_interp_kernels.py]:
```suggestion
@cupy.util.memoize(for_each_device=True)
```

File: cupyx/scipy/ndimage/_interp_kernels.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+        mode=mode,
+        cval=cval,
+        order=order,
+        name='shift',
+        integer_output=integer_output,
+    )
+    return cupy.ElementwiseKernel(in_params, out_params, operation, name)
+
+
+@cupy.util.memoize()
```
Comment (leofang) [inline cupyx/scipy/ndimage/_interp_kernels.py]:
```suggestion
@cupy.util.memoize(for_each_device=True)
```

File: cupyx/scipy/ndimage/_interp_kernels.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
```
Comment (leofang) [inline cupyx/scipy/ndimage/_interp_kernels.py]:
```suggestion
@cupy.util.memoize(for_each_device=True)
```

File: cupyx/scipy/ndimage/_interp_kernels.py [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+        mode=mode,
+        cval=cval,
+        order=order,
+        name='zoom_shift',
+        integer_output=integer_output,
+    )
+    return cupy.ElementwiseKernel(in_params, out_params, operation, name)
+
+
+@cupy.util.memoize()
```
Comment (leofang) [inline cupyx/scipy/ndimage/_interp_kernels.py]:
```suggestion
@cupy.util.memoize(for_each_device=True)
```

Comment (grlee77) [reply cupyx/scipy/ndimage/_interp_kernels.py]: Thanks!

I didn't want to make separate commits for all of these so I did one commit manually and added
`Co-authored-by: Leo Fang <leofang@bnl.gov>` there

Comment (leofang) [reply cupyx/scipy/ndimage/_interp_kernels.py]: Thanks, Gregory. No worries about the coauthor thing. I just thought there's a batch commit button and having all needed changes explicitly laid out might be easier for you.

File: cupyx/scipy/ndimage/_interp_kernels.py (hunk 1 above)
Comment (leofang) [reply cupyx/scipy/ndimage/_interp_kernels.py]: This one is still missed...?

Comment (grlee77) [reply cupyx/scipy/ndimage/_interp_kernels.py]: Whoops... Thanks!

**Comment by grlee77** [comment]
#3184 managed to remove `xshape` from the parameters for `convolve` which would help reduce the number of combinations for `memoize` if a similar thing is possible here.

I did try making indexing changes as in #3179 and most functions stayed about the same, although `map coordinates` did show a substantial improvement. This was prior to the merging of #3191, though which may have given a similar benefit. I will take a look again and see.

**Comment by grlee77** [comment]
I was able to remove `xshape` from the parameterized parameters.

#3191 seems to have made contiguous indexing fast enough that it was not necessarily to make specific indexing changes as done in #3179.

File: cupyx/scipy/ndimage/interpolation.py [hunk 6]
```diff
@@ -1,8 +1,18 @@
-import itertools
 import math
 import warnings
 
 import cupy
+import numpy
+
+from ._interp_kernels import (
```
Comment (asi1024) [inline cupyx/scipy/ndimage/interpolation.py:7]: Importing non-module symbols is prohibited.

File: cupyx/scipy/ndimage/interpolation.py [hunk 7]
```diff
@@ -209,15 +169,15 @@ def affine_transform(input, matrix, offset=0.0, output_shape=None, output=None,
...
-        # TODO(mizuno): Implement zoom_shift
-        matrix = cupy.diag(matrix)
-    elif matrix.shape[0] == matrix.shape[1] - 1:
-        offset = matrix[:, -1]
-        matrix = matrix[:, :-1]
-    elif matrix.shape[0] == input.ndim + 1:
-        offset = matrix[:-1, -1]
-        matrix = matrix[:-1, :-1]
+    if matrix.ndim not in [1, 2]:
+        raise RuntimeError('no proper affine matrix provided')
```
Comment (asi1024) [inline cupyx/scipy/ndimage/interpolation.py:173]: Could you add a test for this exception?

File: cupyx/scipy/ndimage/interpolation.py [hunk 8]
```diff
@@ -233,13 +193,31 @@ def affine_transform(input, matrix, offset=0.0, output_shape=None, output=None,
...
         output_shape = input.shape
 
-    coordinates = cupy.indices(output_shape, dtype=cupy.float64)
-    coordinates = cupy.dot(matrix, coordinates.reshape((input.ndim, -1)))
-    coordinates += cupy.expand_dims(cupy.asarray(offset), -1)
-    ret = _get_output(output, input, output_shape)
-    ret[:] = map_coordinates(input, coordinates, ret.dtype, order, mode, cval,
-                             prefilter).reshape(output_shape)
-    return ret
+    matrix = matrix.astype(float, copy=False)
```
Comment (asi1024) [inline cupyx/scipy/ndimage/interpolation.py:196]: Please use `cupy.float64` for the readability.
```suggestion
    matrix = matrix.astype(cupy.float64, copy=False)
```

Comment (asi1024) [reply cupyx/scipy/ndimage/interpolation.py:196]: BTW, Is this typecast needed? A float64 array looks to be returned without this typecast.

File: cupyx/scipy/ndimage/interpolation.py [hunk 9]
```diff
@@ -295,56 +273,57 @@ def rotate(input, angle, axes=(1, 0), reshape=True, output=None, order=None,
...
+
+    img_shape = numpy.asarray(input_arr.shape)
+    in_plane_shape = img_shape[axes]
+    if reshape:
+        # Compute transformed input bounds
+        iy, ix = in_plane_shape
+        out_bounds = rot_matrix @ [[0, 0, iy, iy],
+                                   [0, ix, 0, ix]]
+        # Compute the shape of the transformed input plane
+        out_plane_shape = (out_bounds.ptp(axis=1) + 0.5).astype(int)
```
Comment (asi1024) [inline cupyx/scipy/ndimage/interpolation.py:304]:
```suggestion
        out_plane_shape = (out_bounds.ptp(axis=1) + 0.5).astype(cupy.int64)
```

File: cupyx/scipy/ndimage/interpolation.py [hunk 10]
```diff
@@ -295,56 +273,57 @@ def rotate(input, angle, axes=(1, 0), reshape=True, output=None, order=None,
...
     if axes[1] < 0:
-        axes[1] += input.ndim
+        axes[1] += input_arr.ndim
     if axes[0] > axes[1]:
-        axes = axes[1], axes[0]
-    if axes[0] < 0 or input.ndim <= axes[1]:
-        raise IndexError
+        axes = [axes[1], axes[0]]
+    if axes[0] < 0 or input_arr.ndim <= axes[1]:
+        raise ValueError('invalid rotation plane specified')
```
Comment (asi1024) [inline cupyx/scipy/ndimage/interpolation.py:285]: Could you add a test for this exception?

File: cupyx/scipy/ndimage/interpolation.py [hunk 11]
```diff
@@ -386,15 +365,37 @@ def shift(input, shift, output=None, order=None, mode='constant', cval=0.0,
...
+            order = 1
+        output = _get_output(output, input)
+        if input.dtype.kind in "iu":
+            input = input.astype(cupy.float32)
+        integer_output = output.dtype.kind in "iu"
+        large_int = _prod(input.shape) > 1 << 31
+        kern = _get_shift_kernel(
+            input.ndim, large_int, input.shape, mode, cval=cval, order=order,
+            integer_output=integer_output)
+        shift = cupy.asarray(shift, dtype=float)
```
Comment (asi1024) [inline cupyx/scipy/ndimage/interpolation.py:396]:
```suggestion
        shift = cupy.asarray(shift, dtype=cupy.float64)
```

File: cupyx/scipy/ndimage/interpolation.py [hunk 12]
```diff
@@ -386,15 +365,37 @@ def shift(input, shift, output=None, order=None, mode='constant', cval=0.0,
...
+            order,
+            mode,
+            cval,
+            prefilter,
+        )
+    else:
+        if order is None:
+            order = 1
+        output = _get_output(output, input)
+        if input.dtype.kind in "iu":
```
Comment (asi1024) [inline cupyx/scipy/ndimage/interpolation.py:389]: Could you use single quotes for our convention?
```suggestion
        if input.dtype.kind in 'iu':
```

File: cupyx/scipy/ndimage/interpolation.py [hunk 13]
```diff
@@ -448,14 +450,37 @@ def zoom(input, zoom, output=None, order=None, mode='constant', cval=0.0,
...
-                            output, order, mode, cval, prefilter)
+        output = _get_output(output, input, shape=output_shape)
+        if input.dtype.kind in "iu":
+            input = input.astype(cupy.float32)
+        integer_output = output.dtype.kind in "iu"
+        large_int = max(_prod(input.shape), _prod(output_shape)) > 1 << 31
+        kern = _get_zoom_kernel(
+            input.ndim, large_int, output_shape, mode, order=order,
+            integer_output=integer_output)
+        zoom = cupy.asarray(zoom, dtype=float)
```
Comment (asi1024) [inline cupyx/scipy/ndimage/interpolation.py:484]:
```suggestion
        zoom = cupy.asarray(zoom, dtype=cupy.float64)
```

**Comment by asi1024** [comment]
@grlee77 Could you check the travis failure?

**Comment by grlee77** [comment]
Okay, Travis is happy now.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
@grlee77 Could you check Jenkins failure?

**Comment by grlee77** [comment]
> @grlee77 Could you check Jenkins failure?

Please try again now. (It looks like SciPy prior to 1.3 raised `IndexError` instead of `ValueError` on out of range axes in `rotate`).

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM. Thanks for your PR!

**Pull request merged**